### DISCUSSION TITLE
Libdatachannel to develop standalone native application

### DISCUSSION THREAD

**Discussion by apvt2019** [paullouisageneau/libdatachannel#578] [General]
Hello,

I want to create an application that can do peer-to-peer media+data communication. I am planning to create an application in C++, that will have a custom GUI application, and no browser involved.
I was looking for a lightweight and easy-to-use library and came across this one. I checked the description of this library and found below mentioned paragraphs.

> libdatachannel is a standalone implementation of WebRTC Data Channels, WebRTC Media Transport, and WebSockets in C++17 with C bindings for POSIX platforms (including GNU/Linux, Android, FreeBSD, Apple macOS and iOS) and Microsoft Windows. WebRTC is a W3C and IETF standard enabling real-time peer-to-peer data and media exchange between two devices.
The library aims at being both straightforward and lightweight with minimal external dependencies, to enable direct connectivity between native applications and web browsers without the pain of importing Google's bloated [reference library](https://webrtc.googlesource.com/src/). The interface consists of somewhat simplified versions of the JavaScript WebRTC and WebSocket APIs present in browsers, in order to ease the design of cross-environment applications.

As far as I understand,
The library is in C++ (what I was looking for).
It is lightweight and has a minimal external dependency (Perfect).

Could you please confirm that if I use this library to develop a standalone native application, does it provide full-fledge media+data functionality including SDP? Does it involve the browser in any manner?
Let me know if I could refer to any such sample application that will help me to understand it better.

**Comment by paullouisageneau** [comment]
Hello,

Your understanding is correct, libdatachannel is precisely made for standalone native applications. It does not involve the browser, but is compatible with the WebRTC implementation present in browsers (a libdatachannel native peer can connect to a JavaScript peer in the browser).

Libdatachannel implements WebRTC Data Channels to send reliable or unreliable data between peers, and WebRTC tracks to send audio and video. It handles all the required SDP negotiation, you just have to provide a signaling channel. However, it is only a network library, it doesn't implement video and audio encoding/decoding by itself and requires an external encoder/decoder (preferably h264).

You should look at the [examples](https://github.com/paullouisageneau/libdatachannel/tree/master/examples) depending of what you want to do. You'll mostly be interested in client (Data Channels), media (media reception from C++), streamer (media sending from C++), and a signaling server.

**Comment by apvt2019** [comment]
Hey Paul,

Thank you for the response.
I checked the example applications and almost of all of them uses browser in one or another way.
In my case, it is going to be completely a native application developed in C++ language. It would be having custom GUI and some business logic, and planning to integrate this library in the application for media+data transfer.

Would you be interested to provide technical consultation on this?
We can arrange a meeting to discuss it.

**Reply by paullouisageneau** [reply]
They use the browser because communicating with browsers tends to be the main use case of WebRTC, however it is not required. 

For instance, you can run two native example clients and connect them together, or write a native sender for the media example, or even write a native receiver for the streamer example.

Sadly I won't have time to provide technical consulting presently, however feel free you can ask for help here or on [Discord](https://discord.gg/jXAP8jp3Nn) or [Gitter](https://gitter.im/libdatachannel/community).

**Comment by murat-dogan** [comment]
Hi,

In my project both ends are native applications. There is no problem, you can use it.

**Reply by apvt2019** [reply]
Hello Murat,

Are you referring to any specific example code in this repository or your personal project?
I have started to go through the "Streamer" example to understand the flow, and to see how many changes would be required in the case of a new GUI.

**Reply by murat-dogan** [reply]
I am talking about my personal project, but it is not at GitHub.
Actually it is straight. Examples in libdatachannel is enough.
